pipeline_server/src/main.rs

```diff
@@ -0,0 +1,873 @@
+//! HTTP server that manages cataloging, compilation,
+//! and execution of SQL programs.
+//!
+//! # API concepts
+//!
+//! * Project.  A project is a SQL script with a non-unique name and a unique ID
+//!   attached to it.  The client can add/remove/modify/compile a project.
+//!   Compilation includes running the SQL-to-DBSP compiler followed by the Rust
+//!   compiler.
+//!
+//! * Configuration.  A project can have multiple configurations associated with
+//!   it.  A configuration specifies input and output streams as a YAML file
+//!   that is deserialized into a `ControllerConfig` object.  Similar to
+//!   projects, one can add/remove/modify configs.
+//!
+//! * Pipeline.  A pipeline is a running instance of a compiled project.  One
+//!   can start multiple pipelines for a project with the same or different
+//!   configs.
+//!
+//! The API is currently single-tenant: there is no concept of users or
+//! permissions.  Multi-tenancy can be implemented by creating a server
+//! instance per tenant, which enables better separation of concerns,
+//! resource isolation and fault isolation compared to buiding multitenancy
+//! into the server.
```

Comment:
what if there's very little traffic to the pipeline?

Reply:
Then it shouldn't use a lot of resources.